src/libcrun/kontain_nonkmexec.c

```diff
@@ -0,0 +1,347 @@
+{
+   execpath_entry_t* epp;
+
+   while (!SLIST_EMPTY(&statep->execpath_head)) {
+      epp = SLIST_FIRST(&statep->execpath_head);
+      SLIST_REMOVE_HEAD(&statep->execpath_head, link);
+      free(epp->path);
+      regfree(&epp->re);
+      free(epp);
+   }
+   memset(&statep->execpath_mtime, 0, sizeof(statep->execpath_mtime));
+}
+
+// Destroy the execpath ok db.
+void
+libcrun_kontain_nonkmexec_clean(void)
+{
+   kontain_execpath_freedb(&execpath_state);
+}
+
+static void
+hash_2_ascii(BYTE hash[SHA256_BLOCK_SIZE], char *ascii)
+{
+   int i;
+   for (i = 0; i < SHA256_BLOCK_SIZE; i++) {
+      snprintf(&ascii[i * 2], 3, "%02x", hash[i]);
+   }
+   ascii[i * 2] = 0;
+}
+
+/*
+ * Compute the sha256 hash of the passed file.
+ * Return the hash in ascii in returned_hash.
+ * Returns:
+ *   0 - success
+ *   != 0 - an errno value describing what failed
+ */
+static int
+sha256_file(char* file, char *returned_hash)
+{
+   int f;
+   unsigned char filebuf[128*1024];
+   SHA256_CTX ctx;
+   BYTE final_hash[SHA256_BLOCK_SIZE];
+   ssize_t bytesread;
+
+   f = open(file, O_RDONLY);
+   if (f < 0) {
+      return errno;
+   }
+
+   sha256_init(&ctx);
+   while ((bytesread = read(f, filebuf, sizeof(filebuf))) > 0) {
+      sha256_update(&ctx, filebuf, bytesread);
+   }
+   if (bytesread < 0) {
+      return errno;
+   }
+   sha256_final(&ctx, final_hash);
+
+   close(f);
+
+   hash_2_ascii(final_hash, returned_hash);
+
+   return 0;
+}
+
+// Split a line from the exec ok file into its fields
+static int
+split_into_fields(char* dbentry, char *fields[])
+{
+   char* saveptr = NULL;
+
+   fields[0] = strtok_r(dbentry, ":", &saveptr);
+   if (fields[0] == NULL) {
+      return EINVAL;
+   }
+   fields[1] = strtok_r(NULL, ":", &saveptr);
+   if (fields[0] == NULL) {
+      return EINVAL;
+   }
+   fields[2] = strtok_r(NULL, ":", &saveptr);
+   if (fields[2] == NULL) {
+      return EINVAL;
+   }
+   return 0;
+}
+
+/*
+ * Read a execpath database and place it into execpath_state_t pointed to by statep.
+ * Entries in the file are a single line with fields separated by colons.
+ * There are 3 fields:
+ *   regular expression to match an input string against
+ *   the file to exec to if the re matches
+ *   a sha of the file being exec'ed to to verify the file is unchanged since the
+ *   db was created.
+ * Failures is this function cause the process to exit.
+ *
+ * We should probably use some sort of json representation for this instead of the
+ * archaic colon separated fields.
+ */
+static void
+kontain_execpath_builddb(execpath_state_t* statep)
+{
+   char buf[1024];
+   FILE* f = NULL;
+   char *fields[3];
+   execpath_entry_t* epp = NULL;
+   execpath_entry_t* tail = NULL;
+   int rc = 0;
+
+   if (statep->execpath_dbpath != NULL) {
+      kontain_execpath_freedb(statep);
+   }
+   f = fopen(statep->execpath_dbpath, "r");
+   if (f == NULL) {
+      libcrun_fail_with_error(errno, "Couldn't open %s", statep->execpath_dbpath);
```

Comment:
isn't it config file ? if it is, failing if  it's not there seems exessive

Reply:
The modularity can be better here.
The caller of this function verifies that the file exists and if it does not, will not call this function.
fopen() failures here mean there is some other problem and we should know about it.
Since crun's logging facilities are so awful, sometimes the only way to find out what really happened is to exit crun.
I really wonder what crun's author was thinking.